src/backends/caffe2/caffe2inputimg.cc

```diff
@@ -19,72 +19,447 @@
+
+    // Create a test db
+    if (_is_testable) {
+
+      if (_uris.size() > 1) {
+	list_images(_uris[1], corresp, corresp_r, test_files, true);
+
+      } else { // Split that data
+	auto it = train_files.begin() + std::floor(train_files.size() * (1.0 - _test_split));
+	test_files.assign(it, train_files.end());
+	train_files.erase(it, train_files.end());
+      }
+      write_images_to_db(_db, test_files);
+    }
+    write_images_to_db(_train_db, train_files);
+
+    // write corresp file
+    std::ofstream correspf(_corresp_file, std::ios::binary);
+    for (auto &kp : corresp) {
+      correspf << kp.first << " " << kp.second << std::endl;
+    }
+    correspf.close();
   }
 
-  int ImgCaffe2InputFileConn::get_tensor_test(caffe2::TensorCPU &tensor, int num) {
+  void ImgCaffe2InputFileConn::write_images_to_db(const std::string &dbname,
+						  const std::vector<std::pair<std::string, int>>
+						  &lfiles) {
+
+    std::unique_ptr<caffe2::db::DB> db(caffe2::db::CreateDB(_db_type, dbname, caffe2::db::NEW));
+
+    // Prefill db entries
+    int chans = channels();
+    int chan_size = _height * _width * sizeof(float);
+    caffe2::TensorProtos protos;
+    caffe2::TensorProto *proto_data = protos.add_protos();
+    proto_data->set_data_type(proto_data->FLOAT);
+    proto_data->mutable_dims()->Add(chans);
+    proto_data->mutable_dims()->Add(_height);
+    proto_data->mutable_dims()->Add(_width);
+    proto_data->mutable_float_data()->Resize(chans * _height * _width, 0);
+    float *ptr = proto_data->mutable_float_data()->mutable_data();
+    std::vector<void *> img(chans);
+    for (int i = 0; i < chans; ++i) {
+      img[i] = ptr;
+      ptr += _height * _width;
+    }
+    caffe2::TensorProto *proto_label = protos.add_protos();
+    proto_label->set_data_type(proto_label->INT32);
+    int &label = *proto_label->mutable_int32_data()->Add();
 
-    int image_count = _images.size();
+    int cv_load = _bw ? CV_LOAD_IMAGE_GRAYSCALE : CV_LOAD_IMAGE_COLOR;
+    cv::Size size(_width, _height);
+
+    // Storing to db
+    int count = 0;
+    const int kMaxKeyLength = 256;
+    char key_cstr[kMaxKeyLength];
+
+    for (const auto &file : lfiles) {
+
+      // Fill db entry
+      std::string out;
+      try {
+	label = file.second;
+	cv::Mat mat = cv::imread(file.first, cv_load);
+	CAFFE_ENFORCE(!mat.empty());
+	cv::resize(mat, mat, size, 0, 0, CV_INTER_CUBIC);
+	mat.convertTo(mat, CV_32F);
+	std::vector<cv::Mat> mats(chans);
+	cv::split(mat, mats);
+	for (int i = 0; i < chans; ++i) {
+	  std::memcpy(img[i], mats[i].data, chan_size);
+	}
+	CAFFE_ENFORCE(protos.SerializeToString(&out));
+      } catch (...) {
+	_logger->warn("Could not load {}", file.first);
+	continue;
+      }
+
+      // Put in db
+      auto txn = db->NewTransaction();
+      int length = snprintf(key_cstr, kMaxKeyLength, "%08d_%s", count, file.first.c_str());
+      txn->Put(std::string(key_cstr, length), out);
+      txn->Commit();
```

Comment:
Don't commit after every addition, as it is too slow. Instead use the 1000 images check and commit there 1000 images at a time. See how it is done in `caffeinputconns.cc`.